datastore/query.go

```diff
@@ -81,6 +84,150 @@ type order struct {
 	Direction sortDirection
 }
 
+// EntityFilter represents a datastore filter
+type EntityFilter interface {
```

Comment:
Instead of making an interface here I wonder if this should just a a builder type or if the Filter should be something that dynamically manipulates a query and returns a query rather than passing in filters to a method on query.

So instead of
```
datastore.NewQuery("foo").FilterEntity(datastore.AND{
    Filters: []EntityFilter{
				PropertyFilter{FieldName: "x", Operator: "==", Value: 4},
				PropertyFilter{FieldName: "y", Operator: "<", Value: 3},
			},
})
```

Something more like:
```
datastore.NewQuery("foo").AndFilter(
	PropertyFilter{FieldName: "x", Operator: "==", Value: 4},
	PropertyFilter{FieldName: "y", Operator: "<", Value: 3},
})
```

Thoughts? @enocom wdyt? I have not really used this client before so I don't know what the current exp is like. Maybe the current design aligns with patterns already in the client.

Reply:
The fluent interface is certainly nice. How would we support complex queries with a mix of And's and Or's?

Otherwise, I think the way @bhshkh has it here is clean and easy to understand (and write sophisticated queries).

Reply:
> The fluent interface is certainly nice. How would we support complex queries with a mix of And's and Or's?

Can you give an example with the current API, I am not sure I mean by complex. Or maybe a test should be added for something like this. I would think that all the filters though just keep returning the query. But like I said I have not used this client so I trust your judgment here if you think the current approach has better support for the features needing to be supported.

Reply:
@codyoss 
Sample complex query : (T = now) and (  ( (J > 97) and (T = tomorrow) ) or (J < 94)  )
https://github.com/googleapis/google-cloud-go/blob/f775d45f95fa7a1ac2d89083c7193c990000abdd/datastore/integration_test.go#L528-L545

Reply:
Ah missed this case sorry, this seems as good in that case.